### PR TITLE
refactor IosProject and MacOSProject to extends XcodeBasedProject to share common

### PR DESCRIPTION
**Pull request by wanglikun7342**
refactor IosProject and MacOSProject to extends XcodeBasedProject to share common, because they have a lot similarity

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/pull/87264

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by wanglikun7342** [comment]
@jmagman please review this pr

**Comment by wanglikun7342** [comment]
@jmagman i think the checks make mistake

File: packages/flutter_tools/lib/src/xcode_project.dart [hunk 1]
```diff
@@ -50,20 +57,132 @@ abstract class XcodeBasedProject {
...
+  }
+  XcodeProjectInfo? _projectInfo;
+
+  Future<Map<String, String>?> _xcodeProjectBuildSettings(XcodeProjectBuildContext buildContext) async {
+    final XcodeProjectInterpreter? xcodeProjectInterpreter = globals.xcodeProjectInterpreter;
+    if (xcodeProjectInterpreter == null || !xcodeProjectInterpreter.isInstalled) {
+      return null;
+    }
+
+    final Map<String, String> buildSettings = await xcodeProjectInterpreter.getBuildSettings(
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/xcode_project.dart:169]: I don't think `XcodeProjectInterpreter.getBuildSettings()` is set up to match how macOS builds.  Ideally they would be shared, but I believe the macOS build code was kind of thrown together until it worked and no one went back to use `XcodeProjectInterpreter` in macOS build.

For example
https://github.com/flutter/flutter/blob/089323487fe1b5cc402bbe65c47383f61eba241c/packages/flutter_tools/lib/src/ios/xcodeproj.dart#L187
vs
https://github.com/flutter/flutter/blob/db930ba32f320cb0ca53595a2df0e4a9fe15101d/packages/flutter_tools/lib/src/macos/build_macos.dart#L104-L106

Though in practice it probably doesn't matter for `PRODUCT_BUNDLE_IDENTIFIER`.

**Review by jmagman — CHANGES_REQUESTED** [review]
What problem exactly are you trying to solve?  I don't think the macOS bundle ID is used anywhere, so while this is more correct than the placeholder bundle ID, it's adding seconds to the macOS build without actually adding any functionality.

From https://github.com/flutter/flutter/pull/87264#discussion_r682274575
> Which seems like a lot since the `BuildableMacOSApp` id isn't used. I'm actually okay with this staying a placeholder hack until it's actually used somewhere.

I like the `abstract class XcodeBasedProject extends FlutterProjectPlatform` refactoring work to share more logic between macOS and iOS and would happy to review just that change.

**Comment by wanglikun7342** [comment]
@jmagman I have changed the pr,please review again

**Comment by wanglikun7342** [comment]
@jmagman this pr need your review again

**Review by jmagman — COMMENTED** [review]
This refactor looks good to me.  Can you update the title and description with what it's now doing, and we can merge it.

**Comment by wanglikun7342** [comment]
@Hixie this pr need a test exemption

**Comment by Hixie** [comment]
test-exempt: code refactor with no semantic change

**Comment by wanglikun7342** [comment]
@jmagman done

**Review by jmagman — APPROVED** [review]
LGTM, thanks!

**Pull request merged**